core/lib/state/src/rocksdb/mod.rs

```diff
@@ -131,23 +152,157 @@ impl RocksdbStorage {
         tracing::info!(
             "Secondary storage for L1 batch #{latest_l1_batch_number} initialized, size is {estimated_size}"
         );
+
+        self.save_missing_enum_indices(conn).await;
+    }
+
+    async fn apply_storage_logs(
+        &mut self,
+        storage_logs: HashMap<StorageKey, H256>,
+        conn: &mut StorageProcessor<'_>,
+    ) {
+        let (logs_with_known_indices, logs_with_unknown_indices): (Vec<_>, Vec<_>) = self
+            .process_transaction_logs(storage_logs)
+            .into_iter()
+            .partition_map(|(key, (value, index))| match index {
+                Some(index) => Either::Left((key, (value, index))),
+                None => Either::Right((key, value)),
+            });
+        let keys_with_unknown_indices: Vec<_> = logs_with_unknown_indices
+            .iter()
+            .map(|(key, _)| key.hashed_key())
+            .collect();
+
+        let enum_indices = conn
+            .storage_logs_dedup_dal()
+            .enum_indices_for_keys(&keys_with_unknown_indices)
+            .await;
+        assert_eq!(keys_with_unknown_indices.len(), enum_indices.len());
+        self.pending_patch.state = logs_with_known_indices
+            .into_iter()
+            .chain(
+                logs_with_unknown_indices
+                    .into_iter()
+                    .zip(enum_indices)
+                    .map(|((key, value), index)| (key, (value, index))),
+            )
+            .collect();
+    }
+
+    async fn save_missing_enum_indices(&self, conn: &mut StorageProcessor<'_>) {
+        let (Some(start_from), true) = (
+            self.enum_migration_start_from(),
+            self.enum_index_migration_chunk_size > 0,
+        ) else {
+            return;
+        };
+
+        let started_at = Instant::now();
+        let progress_percentage = BigDecimal::from(100)
+            * u256_to_big_decimal(h256_to_u256(start_from))
+            / u256_to_big_decimal(U256::MAX);
+        tracing::info!(
+            "RocksDB enum index migration is not finished, current progress {}%, starting from key {start_from:?}",
+            progress_percentage.with_prec(5)
+        );
+
+        let mut write_batch = self.db.new_write_batch();
+        let (keys, values): (Vec<_>, Vec<_>) = self
+            .db
+            .from_iterator_cf(StateKeeperColumnFamily::State, start_from.as_bytes())
+            .take(self.enum_index_migration_chunk_size)
+            .filter_map(|(key, value)| {
+                (key.as_ref() != Self::BLOCK_NUMBER_KEY
+                    && key.as_ref() != Self::ENUM_INDEX_MIGRATION_PROGRESS_KEY
+                    && key.as_ref() != Self::ENUM_INDEX_MIGRATION_FINISHED
+                    && value.len() != 40)
+                    .then(|| (H256::from_slice(&key), H256::from_slice(&value[..32])))
```

Comment:
I'd suggest to create a type for values and encoding / decoding methods for it in order to improve readability:

```rust
#[derive(Debug, Clone, Copy)]
struct StateValue {
    value: H256,
    enum_index: Option<u64>,
}

impl StateValue {
    fn deserialize(bytes: &[u8]) -> Self { /* ... */ }
    fn serialize(&self) -> Vec<u8> { /* ... */ }
}
```